src/Data/Text/Internal/Search.hs

```diff
@@ -54,14 +54,19 @@ data T = {-# UNPACK #-} !Word64 :* {-# UNPACK #-} !Int
 indices :: Text                -- ^ Substring to search for (@needle@)
         -> Text                -- ^ Text to search in (@haystack@)
         -> [Int]
-indices (Text narr noff nlen)
+indices needle@(Text narr noff nlen)
   | nlen == 1 = scanOne (A.unsafeIndex narr noff)
   | nlen <= 0 = const []
-  | otherwise = scan
+  | otherwise = indices' needle
+{-# INLINE indices #-}
+
+-- | nlen must be >= 2, otherwise nindex causes access violation
```

Comment:
https://github.com/haskell/text/blob/922d9cc7c537d55e3be2dba2343ef43e25493ace/src/Data/Text/Internal/Search.hs#L57-L64

When `nlen = noff = 0` we have `z = A.unsafeIndex narr (-1)`. We never use `z` when `nlen = 0`, but since it is a strict binding, GHC evaluates it any way: https://github.com/haskell/text/runs/4275761397?check_suite_focus=true#step:6:658

In practice access violation is likely to be invisible, because it reads bytes somewhere from `ByteArray` header. The issue is caught only in `developer` build with bound checks enabled.

Reply:
I was wondering about the `nlen = 1` case, is that just because we can efficiently special case?

Reply:
`nlen = 1` explodes with access violation for similar reasons in 
https://github.com/haskell/text/blob/922d9cc7c537d55e3be2dba2343ef43e25493ace/src/Data/Text/Internal/Search.hs#L71

Yes, `nlen = 1` is a special case, allowing us to use `memchr`, which is a highly optimized routine.